lib/src/fixed_datetime_codec.dart

```diff
@@ -0,0 +1,153 @@
+// Copyright (c) 2022, the Dart project authors.  Please see the AUTHORS file
+// for details. All rights reserved. Use of this source code is governed by a
+// BSD-style license that can be found in the LICENSE file.
+
+/// A class for parsing and formatting dates for a fixed format string. For
+/// example, calling
+/// `FixedDateTimeCodec('YYYYMMDDhhmmss').decodeToLocal('19960425050322')` has
+/// the same result as calling `DateTime(1996, 4, 25, 5, 3, 22)`.
+///
+/// The allowed characters are
+/// * Y	a digit used in the time scale component “calendar year”
+/// * M	a digit used in the time scale component “calendar month”
+/// * D	a digit used in the time scale component “calendar day”
+/// * E	a digit used in the time scale component “decade”
+/// * C	a digit used in the time scale component “century”
+/// * h	a digit used in the time scale component “clock hour”
+/// * m	a digit used in the time scale component “clock minute”
+/// * s	a digit used in the time scale component “clock second”
+/// as specified in the ISO 8601 standard.
+///
+/// Non-allowed characters in the format [_pattern] are ignored, therefore
+/// `YYYY kiwi MM` is the same format string as `YYYY------MM`.
+///
+/// Note: this class differs from [DateFormat] in that here, the characters are
+/// treated literally, i.e. the format string `YYY` matching `996` would result in
+/// the same as calling `DateTime(996)`. [DateFormat] on the other hand uses the
+/// specification in https://www.unicode.org/reports/tr35/tr35-dates.html#Date_Field_Symbol_Table,
+/// the format string (or "skeleton") `YYY` specifies only the padding, so
+/// `1996` would be a valid match. This limits is use to format strings
+/// containing delimiters, as the parser would not know how many digits to take
+/// otherwise.
+///
+/// Also, this parser does not know about locales and parses to the current
+/// locale only.
+class FixedDateTimeCodec {
+  static const _validChars = ['Y', 'M', 'D', 'E', 'C', 'h', 'm', 's'];
+
+  final String _pattern;
+  final _occurences = <String, _Range>{};
+
+  FixedDateTimeCodec(this._pattern) {
+    String current = '';
+    for (var i = 0; i < _pattern.length; i++) {
+      var char = _pattern[i];
+      if (!_validChars.contains(char)) {
+        current = '';
+        continue;
+      }
+      var newChar = current != char;
+      if (newChar) {
+        var hasNotSeenBefore = !_occurences.containsKey(char);
+        if (hasNotSeenBefore) {
+          _occurences[char] = _Range(i, i);
+        } else {
+          throw Exception(
+              'The _pattern string "$_pattern" contains multiple instances of the formatting char block $char, both at position ${_occurences[char]!.from} and $i');
+        }
+      }
+      current = char;
+      _occurences.update(current, (value) => _Range(value.from, value.to + 1));
+    }
+  }
+
+  /// Convert a datetime to a string exactly as specified by the [_pattern].
+  String encode(DateTime dt) {
```

Comment:
Avoid abbrevations, so `(DateTime dateTime)` or `(DateTime timeStamp)`.
Probably the former.